### ISSUE TITLE
args hash contains a key which wasn't part of mutation variables

### ISSUE THREAD

**Issue by bgentry** [rmosolgo/graphql-ruby#472]
As [requested](https://graphql.slack.com/archives/ruby/p1483405917000724) by @rmosolgo, opening a bug about this.

In my app, I have an update form. I have that form set up to submit a mutation with `variables` consisting of only the atttributes that have actually changed. Here's a simple example of that query, though you could imagine it having many more arguments:

```graphql
mutation updateObject($id: IDType!, $name: String, $enabled: Boolean) {
  updateObject(id: $id, name: $name, enabled: $enabled) {
    ...
  }
}
```

If I toggle `enabled` on my form, this is the JSON body of the mutation request:

```json
{"query":"mutation updateWatcher($id: IDType!, $name: String, $enabled: Boolean) {
  updateWatcher(id: $id, name: $name, enabled: $enabled) {
    watcher {
      id
      name
      enabled
      __typename
    }
    __typename
  }
}
","variables":{"enabled":true,"id":"27tlpoPeXm6odAxj3paGQP"},"operationName":"updateWatcher"}
```

In particular, note that `variables` is `{"enabled":true,"id":"27tlpoPeXm6odAxj3paGQP"}`. The `name` attr was not changed, and thus was not provided in the `variables`.

However, on my graphql-ruby server, this same request shows the following args:

```ruby
[2] pry(#<UpdateWatcherResolver>)> args.to_h
=> {"id"=>"27tlpoPeXm6odAxj3paGQP", "name"=>nil, "schedule_every"=>60}
[3] pry(#<UpdateWatcherResolver>)> args.key?("name")
=> true
[4] pry(#<UpdateWatcherResolver>)> args.key?(:name)
=> true
[5] pry(#<UpdateWatcherResolver>)> args.class
=> GraphQL::Query::Arguments
```

Note the `"name" => nil` bit. I did not provide a `name` key in my `variables`, but graphql-ruby added it to the args hash, presumably because the variable was referenced in the mutation itself.

I wasn't sure if this was the correct behavior, and @rmosolgo seems to think it might be a bug.

What I'd like to be able to do is have a single statically-defined query in the frontend that includes all the variables I *might* update for that object on that specific mutation. I'd like that one query to work as a PATCH-style update for any valid combination of `variables` and have my backend only see the vars that I actually provided as `variables`.

I'm going to try using an input object for now, and hopefully that gets me the behavior I want.

**Comment by bgentry** [comment]
An update: using an input object does give me exactly the behavior I want (variables whose keys are not provided do not show up at all in the args hash). Probably doesn't make sense that this behavior differs between root-level arguments and input objects, so the bug label seems appropriate :v:

**Comment by cjoudrey** [comment]
I'll take a stab at fixing this.

@bgentry @rmosolgo here's a quick repro script just to confirm I understand this correctly.

**Schema**

```ruby
require 'graphql'

QueryType = GraphQL::ObjectType.define do
  name "Query"

  field :hello, types.String do
    argument :test, types.String
    resolve -> (_, args, _) {
      p args.to_h
      "world"
    }
  end
end

Schema = GraphQL::Schema.define do
  query QueryType
end
```

**Case 1 (bug)**

```ruby
Schema.execute('''
  query ($test: String) {
    hello(test: $test)
  }
''')

# Actual:
# {"test"=>nil}

# Expected:
# {}
```

**Case 2 (okay 👌)**
```ruby
Schema.execute('''
  query ($test: String) {
    hello(test: $test)
  }
''', variables: { 'test' => 'present' })

# Actual:
# {"test"=>"present"}

# Expected:
# {"test"=>"present"}
```

**Case 3 (okay 👌)**

```ruby
Schema.execute('''
  query ($test: String) {
    hello(test: $test)
  }
''', variables: { 'test' => nil })

# Actual:
# {"test"=>nil}

# Expected:
# {"test"=>nil}
```

**Comment by bgentry** [comment]
@cjoudrey yep, looks correct! :v:/

**Comment by rmosolgo** [comment]
Taking a try at this in #479, does that seem ok?

**Comment by bgentry** [comment]
Unfortunately, on 1.4.1, I'm still seeing this. Here's my query:

```graphql

        mutation createSomething(
          $value: Boolean!,
          $duration: Int
        ) {
          createSomething(
            value: $value,
            duration: $duration
          ) {
            something {
              id
            }
          }
        }
```

I'm passing it the following variables:

```ruby
{ :value => false }
```

In my resolver, I'm seeing the following args:

```ruby
{ "value" => false, "duration" => nil }
```

Not that `duration` was not present in the `variables` hash, but its key did show up in the resolver's args. This is still different than how it behaves with an `InputObjectType`. In that case, you can leave out attrs on the input object and they will not show up at all in the resulting args.

Is this caused by the fact that I'm referencing a variable `$duration` in my query that is not actually present in the variables?

**Comment by rmosolgo** [comment]
Just released 1.4.2 which includes another attempt to fix this, could you give it a try?
